pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -157,78 +136,92 @@ def circuit(weights):
     """Computes the metric tensor of a tape using the adjoint method and a given device."""
     # pylint: disable=protected-access
-    if device.shots is not None:
+    if tape.shots:
         raise ValueError(
             "The adjoint method for the metric tensor is only implemented for shots=None"
         )
+
+    if set(tape.wires) != set(range(tape.num_wires)):
+        wire_map = {w: i for i, w in enumerate(tape.wires)}
+        tape = qml.map_wires(tape, wire_map)
     tape = qml.transforms.expand_trainable_multipar(tape)
 
     # Divide all operations of a tape into trainable operations and blocks
     # of untrainable operations after each trainable one.
     trainable_operations, group_after_trainable_op = _group_operations(tape)
 
-    dim = 2**device.num_wires
+    dim = 2**tape.num_wires
     # generate and extract initial state
-    psi = device._create_basis_state(0)
+    prep = tape[0] if len(tape) > 0 and isinstance(tape[0], qml.operation.StatePrep) else None
+
+    interface = qml.math.get_interface(*tape.get_parameters(trainable_only=False))
+    psi = qml.devices.qubit.create_initial_state(tape.wires, prep, like=interface)
 
     # initialize metric tensor components (which all will be real-valued)
     like_real = qml.math.real(psi[0])
     L = qml.math.convert_like(qml.math.zeros((tape.num_params, tape.num_params)), like_real)
     T = qml.math.convert_like(qml.math.zeros((tape.num_params,)), like_real)
 
-    psi = _apply_operations(psi, group_after_trainable_op[-1], device)
+    for op in group_after_trainable_op[-1]:
+        psi = qml.devices.qubit.apply_operation(op, psi)
 
     for j, outer_op in enumerate(trainable_operations):
         generator_1, prefactor_1 = qml.generator(outer_op)
-        generator_1 = qml.matrix(generator_1)
+        generator_unitary = _get_op_for_generator(generator_1, like_real)
 
         # the state vector phi is missing a factor of 1j * prefactor_1
-        phi = device._apply_unitary(
-            psi, qml.math.convert_like(generator_1, like_real), outer_op.wires
-        )
+        phi = qml.devices.qubit.apply_operation(generator_unitary, psi)
 
-        phi_real = qml.math.reshape(qml.math.real(phi), (dim,))
-        phi_imag = qml.math.reshape(qml.math.imag(phi), (dim,))
+        phi_real, phi_imag = _reshape_real_imag(phi, dim)
         diag_value = prefactor_1**2 * (
             qml.math.dot(phi_real, phi_real) + qml.math.dot(phi_imag, phi_imag)
         )
-        L = qml.math.scatter_element_add(L, (j, j), diag_value)
 
         lam = psi * 1.0
+        lam_real, lam_imag = _reshape_real_imag(lam, dim)
         lam_real = qml.math.reshape(qml.math.real(lam), (dim,))
         lam_imag = qml.math.reshape(qml.math.imag(lam), (dim,))
 
         # this entry is missing a factor of 1j
         value = prefactor_1 * (qml.math.dot(lam_real, phi_real) + qml.math.dot(lam_imag, phi_imag))
+        L = qml.math.scatter_element_add(L, (j, j), diag_value)
```

Comment:
Wouldn't it be good to keep this close to where `diag_value` is computed? :thinking:

Reply:
Should be fine. I think this is just a byproduct of trying and failing to extract more helper methods.